### PR TITLE
+str Add Source#mergePrioritizedN.

### PR DESCRIPTION
**Pull request by He-Pin**
References https://github.com/akka/akka/issues/31131

### CODE REVIEW

File: akka-docs/src/main/paradox/stream/operators/Source/mergePrioritizedN.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+## Reactive Streams semantics
+
+@@@div { .callout }
+
+**emits** when one of the inputs has an element available, preferring inputs based on their priorities if multiple have elements available
+
+**backpressures** when downstream backpressures
+
+**completes** when all upstreams complete (This behavior is changeable to completing when any upstream completes by setting `eagerComplete=true`.)
```
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/operators/Source/mergePrioritizedN.md:34]:
```suggestion
**completes** when all upstreams complete (or when any upstream completes if `eagerComplete=true`.)
```

File: akka-stream/src/main/scala/akka/stream/impl/Stages.scala [hunk 2]
```diff
@@ -76,6 +76,7 @@ import akka.stream.Attributes._
     val merge = name("merge")
     val mergePreferred = name("mergePreferred")
     val mergePrioritized = name("mergePrioritized")
+    val mergePrioritizedN = name("mergePrioritizedN")
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Stages.scala:79]: I think the default `mergePrioritized` name (set in `MergePrioritized`) is enough and that this is just a convenience factory, so skip this.

File: akka-docs/src/main/paradox/stream/operators/Source/mergePrioritizedN.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+@@signature [Source.scala](/akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala) { #mergePrioritized }
+
+## Description
+
+Merge multiple sources. Prefer sources depending on priorities if all sources have elements ready. If a subset of all
+sources have elements ready the relative priorities for those sources are used to prioritize. For example, when used 
+with only three sources sourceA sourceB and sourceC, the sourceA has a probability of `(priorityOfA) / (priorityOfA + priorityOfB + priorityOfC)` of being 
+prioritized and similarly for the rest source. The priorities for each source must be positive integers, and mast have 
+the same order of sources.
```
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/operators/Source/mergePrioritizedN.md:17]:
```suggestion
with only three sources `sourceA`, `sourceB` and `sourceC`, the `sourceA` has a probability of `(priorityOfA) / (priorityOfA + priorityOfB + priorityOfC)` of being 
prioritized and similarly for the rest of the sources. The priorities for each source must be positive integers, and mast have 
the same order of sources.
```

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 4]
```diff
@@ -974,4 +974,29 @@ object Source {
...
+   * '''backpressures''' when downstream backpressures
+   *
+   * '''completes''' when both upstreams complete (This behavior is changeable to completing when any upstream completes by setting `eagerComplete=true`.)
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def mergePrioritizedN[T](
+      sources: immutable.Seq[Source[T, _]],
+      priorities: immutable.Seq[Int],
+      eagerComplete: Boolean = false): Source[T, NotUsed] = {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:992]: I wonder if `sourcesAndPriorities: immutable.Seq[(Source[T, _], Int))]` wouldn't be a better API than two seqs that need to be in sync, with a single seq there is no way you can accidentally mismatch the number of elements and it is also obvious which source the priority is for.

File: akka-docs/src/main/paradox/stream/operators/Source/mergePrioritizedN.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+@@signature [Source.scala](/akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala) { #mergePrioritized }
+
+## Description
+
+Merge multiple sources. Prefer sources depending on priorities if all sources have elements ready. If a subset of all
+sources have elements ready the relative priorities for those sources are used to prioritize. For example, when used 
+with only three sources `sourceA`, `sourceB` and `sourceC`, the `sourceA` has a probability of `(priorityOfA) / (priorityOfA + priorityOfB + priorityOfC)` of being 
+prioritized and similarly for the rest of the sources. The priorities for each source must be positive integers, and mast have 
+the same order of sources.
```
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/operators/Source/mergePrioritizedN.md:17]:
```suggestion
prioritized and similarly for the rest of the sources. The priorities for each source must be positive integers.
```

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 6]
```diff
@@ -974,4 +974,27 @@ object Source {
...
+   *
+   * '''backpressures''' when downstream backpressures
+   *
+   * '''completes''' when both upstreams complete (This behavior is changeable to completing when any upstream completes by setting `eagerComplete=true`.)
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def mergePrioritizedN[T](
+      sourcesAndPriorities: immutable.Seq[(Source[T, _], Int)],
+      eagerComplete: Boolean = false): Source[T, NotUsed] = {
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:991]: Default values makes keeping binary compatibility problematic, so let's skip that, if we need a default shortcut an overload with only one parameter is better.
```suggestion
      eagerComplete: Boolean): Source[T, NotUsed] = {
```

**Review by johanandren — COMMENTED** [review]
Two more small things, then I think this is good to go.

**Comment by hepin1989** [comment]
@johanandren Thanks for the CR

**Review by johanandren — APPROVED** [review]

**Comment by hepin1989** [comment]
@patriknw Do you have a time to take a look at it, I want to make use of it in next release ,thanks.

**Review by patriknw — APPROVED** [review]
LGTM, thanks

**Pull request merged**